docs/fp-core.md

```diff
@@ -0,0 +1,90 @@
+# Finality Provider Core
+
+This doc is to provide basic invariants and heuristics one should follow to
+build the finality provider. This part should be agnostic to any consumer
+chains.
+
+## Status Transitions
+
+The lifecycle of a finality provider instance involves 5 status: `Registered`,
+`Inactive`, `Active`, `Jailed`, `Slashed`. The status transition is depicted
+in the following diagram.
+
+![Finality Provider Status Transition](./fp-status-transition.png).
+
+## Invariants
+
+We define invariants that the finality provider should satisfy:
+
+- Finality signature should be submitted for a block if:
+  - the block is not finalized,
+  - the finality provider has positive voting power for this block,
+  - the finality provider has not cast a finality signature for this block, and
+  - the finality provider has cast finality votes for all non-finalized blocks
+    prior to this block for which it has voting power.
+
+- Public randomness should be committed for a block if:
+  - the block is not finalized, and
+  - the finality provider has not committed any randomness for this block
+
+- A finality provider instance should not be running if it’s status is `slashed`
+  or `jailed` on Babylon.
+
+#### Internal variables
+
+The finality provider maintains the following internal variables.
+* `last_voted_height`: the last block height on which the fp has cast a vote,
+* `last_processed_height`: the last block height on which the fp either has
+  cast a vote or it does not have voting power, and
+* `next_height`: the next block to poll.
```

Comment:
Do you think we should distinguish between internal variables of the FP instance and the poller?

Reply:
As there are not many internal variables to maintain, I think we can just list all of them here.